### PR TITLE
[LLVM][NVPTX]: Add intrinsic for setmaxnreg

### PR DESCRIPTION
**Pull request by durga4github**
This patch adds an intrinsic for setmaxnreg PTX instruction.
* PTX Doc link for this instruction: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#miscellaneous-instructions-setmaxnreg

* The i32 argument, an immediate value, specifies the actual
  absolute register count for the instruction.
* The `setmaxnreg` instruction is available in SM90a.
  So, this patch adds 'hasSM90a' predicate to use in
  the NVPTX backend.
* lit tests are added to verify the lowering of the intrinsic.
* Verifier logic (and tests) are added to test the register
  count range and divisibility-by-8 requirements.

### CODE REVIEW

**Comment by durga4github** [comment]
@grypp , @joker-eph , Could you please help with the review?

File: llvm/include/llvm/IR/IntrinsicsNVVM.td
Comment (joker-eph) [inline llvm/include/llvm/IR/IntrinsicsNVVM.td:4716]: Why is this intrinsic convergent?

File: llvm/include/llvm/Support/NVVMIntrinsicFlags.h
Comment (joker-eph) [inline llvm/include/llvm/Support/NVVMIntrinsicFlags.h:1]: It does not seem usual to have Target-specific intrinsic info in libSupport, are there precedents?

File: llvm/include/llvm/IR/IntrinsicsNVVM.td
Comment (durga4github) [reply llvm/include/llvm/IR/IntrinsicsNVVM.td:4716]: The PTX instruction has the '.aligned' semantics i.e. all threads must execute the same setmaxnreg instruction.

Quoting from the PTX spec:
--
"The mandatory .aligned qualifier indicates that all threads in the warpgroup must execute the same setmaxnreg instruction. In conditionally executed code, setmaxnreg instruction should only be used if it is known that all threads in warpgroup evaluate the condition identically, otherwise the behavior is undefined."

File: llvm/include/llvm/Support/NVVMIntrinsicFlags.h
Comment (durga4github) [reply llvm/include/llvm/Support/NVVMIntrinsicFlags.h:1]: I could not find exact precedents.

As I mentioned in the commit message, we would like to use these enums/flags even in upstream clients (i.e. producers of the intrinsic like NVVM Dialect for example) and not just in the NVPTX backend.  So, I tried to choose a location that looks relevant/compatible both upstream and downstream.

With that, I found a few *Metadata.h, *Attributes.h files in this directory and thought it might be an appropriate one. But, I am open to other preferences that we may have.

Comment (joker-eph) [reply llvm/include/llvm/Support/NVVMIntrinsicFlags.h:1]: Can this be in the IR directory maybe?

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (grypp) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:169]: using `901` safe right? I mean if we have sm_91, it would not be a problem?

File: llvm/include/llvm/IR/IntrinsicsNVVM.td
Comment (grypp) [reply llvm/include/llvm/IR/IntrinsicsNVVM.td:4716]: Does "IntrConvergent" means convergence within a warpgroup? This instruction helps establish smaller register sizes for first warpgroup and larger register sizes for second warpgroup. At least this is how we use for warp specilized GEMM kernels

**Review by grypp — COMMENTED** [review]
Looks good to me in general.

PTX doc has restriction for this instruction. Is it possible to check that?

```
Operand imm-reg-count is an integer constant. The value of imm-reg-count must be in the range 24 to 256 (both inclusive) and must be a multiple of 8.
```

File: llvm/test/CodeGen/NVPTX/setmaxnreg.ll
Comment (grypp) [inline llvm/test/CodeGen/NVPTX/setmaxnreg.ll:12]: Do we want to prefer two intrinsic for inc/dec? 
```
llvm.nvvm.setmaxnreg.inc.sync.aligned.u32
llvm.nvvm.setmaxnreg.dec.sync.aligned.u32
```

File: llvm/include/llvm/IR/IntrinsicsNVVM.td
Comment (durga4github) [reply llvm/include/llvm/IR/IntrinsicsNVVM.td:4716]: Yes,

File: llvm/test/CodeGen/NVPTX/setmaxnreg.ll
Comment (durga4github) [reply llvm/test/CodeGen/NVPTX/setmaxnreg.ll:12]: This is what I initially thought, but this would leave us with no option to extend the intrinsic in future (may be with different action settings).

If you think, we should add two explicit intrinsics (since this is just two and not a lot more now), I am fine with that too. Let me know what you think,

Also, I will add verifier logic for the %8 and range checks.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:168]: `getSmVersion()` is redundant here as `getFullSmVersion()` provides a more specific version comparison.

On a side note, now that we have to deal with `sm_??a` variants we may provide a generic `hasSMA<N>`.
`hasSM90a` here is fine for now.

File: llvm/test/CodeGen/NVPTX/setmaxnreg.ll
Comment (Artem-B) [reply llvm/test/CodeGen/NVPTX/setmaxnreg.ll:12]: I think separate intrinsics make more sense here. The `inc/dec` flags translate into a PTX instruction *name*, so matching intrinsic name would be a better fit, and has a bonus of simplifying the implementation.

File: llvm/include/llvm/Support/NVVMIntrinsicFlags.h
Comment (jholewinski) [reply llvm/include/llvm/Support/NVVMIntrinsicFlags.h:1]: IR feels appropriate, as this header contains definitions for intrinsic operand encoding.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (durga4github) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Fixed to use only getFullSmVersion() in the latest revision.

File: llvm/test/CodeGen/NVPTX/setmaxnreg.ll
Comment (durga4github) [reply llvm/test/CodeGen/NVPTX/setmaxnreg.ll]: Sure, updated in the latest revision,

File: llvm/include/llvm/Support/NVVMIntrinsicFlags.h
Comment (durga4github) [reply llvm/include/llvm/Support/NVVMIntrinsicFlags.h]: Sure, Thanks both of you for the direction. I will use IR directory for other intrinsics.
(For this intrinsic, we are not using flags in the latest revision.)

Resolving this,

**Review by grypp — APPROVED** [review]
Thank you for addressing the comments. It looks good to me, but let's wait for others to review it.

File: llvm/lib/IR/Verifier.cpp
Comment (Artem-B) [inline llvm/lib/IR/Verifier.cpp:6040]: Nit. "must be in multiples of 8" may work a bit better.

After all all integers are divisible by 8, it's just the remainder for some is non-zero. :-)

**Review by Artem-B — APPROVED** [review]
LGTM with a small nit.

File: llvm/lib/IR/Verifier.cpp
Comment (durga4github) [reply llvm/lib/IR/Verifier.cpp]: Sure, totally agree with you ;-)

**Comment by durga4github** [comment]
@Artem-B , I don't have commit access. Please help merge the commit.

**Pull request merged**
